### ISSUE TITLE
Internationalize Hardcoded Copy

### ISSUE THREAD

**Issue by benhalpern** [forem/forem#14888]
### This is related to the [Forem Internationalization effort outlined here](https://dev.to/devteam/forem-hacktoberfest-let-s-internationalize-404n).

**Describe the bug**

This does not refer to any one specific area, but the request to look around the code and find an opportunity to translate anything that is hardcoded into the app.

Read the linked post above for more details.

Here is an example pull request for this task: https://github.com/forem/forem/pull/14884

You do not need to be assigned to any one in particular. Please read the post, and all documentation to get the full context into what the expectations are. Your request will be related to this issue, but will not close it.

**Please read all material carefully and keep your pull requests clean and small.**

Happy coding!

**Comment by papuruth** [comment]
Hey @benhalpern Can I work on this?

**Comment by berviantoleo** [comment]
Cool.

I have some questions.

* How about the link? I think currently the link will navigate to the default (which is english), instead of `locale/:locale/{another-link}`.
* How I handle the singular/plural things? Just make different entries?

Thank you.

**Comment by metamoni** [comment]
It's great to have this as an open ticket, but it might be worth breaking it down into areas/flows, to make sure people don't translate the same files. What do you think?

**Comment by metamoni** [comment]
For those of you who aren't familiar with it, the [Microsoft Terminology Collection](https://www.microsoft.com/en-us/language) is a great resource for this type of work 🙂

**Comment by yheuhtozr** [comment]
Could you give me some idea when to use this `core` prefix? Does it mean what typically appear to the top page?
https://github.com/forem/forem/blob/e5226b9951e1f37474f92192d9e0fc9f769e14e1/config/locales/en.yml#L3

Because I have a lot of strings to contribute, with [such a hierarchy](https://github.com/yheuhtozr/forem/tree/forked/app/i18n/views) in my instance (or maybe more browsable in [the previous monolithic locale file](https://github.com/yheuhtozr/forem/blob/7df70c9f29e4bdd789c04495ab807e34b38ed7a8/app/i18n/en-US.yml)). Most ones only appear in a single or a few related templates under each category. Only `follow*` and `report_abuse` among the current entries seem to truly span multiple categories (though I still recommend separating "Follow" strings based on what is followed).

**Comment by rhymes** [comment]
@yheuhtozr thank you!!

**Comment by citizen428** [comment]
Yes, let's please get all the views merged first. The current frequency of submissions is a bit hard to keep up with.

**Comment by yheuhtozr** [comment]
Now presuming my last batch from ERB to be accepted, what will be the next step?

- i18n for other parts of Ruby code
- i18n for asset pipeline JS code base
- i18n for Preact code base
- locale-aware refactoring of ERBs (cache, links, etc)

In the meantime, are we going to accept locale file contributions once the views i18n is complete? What are the remaining obstacle(s) preventing new locales coming in?

**Comment by citizen428** [comment]
@yheuhtozr Just my personal preference but I think `i18n for other parts of Ruby code` sounds good.

**Comment by yheuhtozr** [comment]
My next batches on Ruby part i18n will be ready early next month.

**Comment by citizen428** [comment]
Hm, @Ridhwana how do you feel about brining I18n to admin at this point?

**Comment by yheuhtozr** [comment]
Caveat is that I didn't do much i18n work in the admin realm so it will be likely to start from scratch or call for other people who already did.

**Comment by Ridhwana** [comment]
> Hm, @Ridhwana how do you feel about brining I18n to admin at this point?

I think we can hold off on the admin for the moment, we're busy reworking most of the views, which means that it's prone to big changes. As we update the views, then we can start working on them.

**Comment by yheuhtozr** [comment]
@citizen428 @nickytonline

So, here are two architectural concerns about JS (string) i18n as far as I tried:

- Much [as I stated](https://github.com/forem/forem/pull/14932#discussion_r722095473), the current provisional implementation with `i18n-js` seems suboptimal in the long run, of which the greatest weakness is that you have to reinvent many things that JS `Intl` already warrants for each language (plural, units, dates...). I think moving into another more popular JS library would be better, but it'll also mean we cannot share i18n in YAMLs directly (while I think okay because actual shared strings are small in number). What do you think about it?
- Not sure what is the best practice of loading i18n entries:
  In my current implementation not very different from [this PR](https://github.com/forem/forem/pull/14932), you can see each JSON can take <= 20KB per language; this could be problematic if loaded at once when we support a lot of languages, so ideally we should only load one according to the viewer's locale (even better if the relevant portion only), but:
  - in the Sprockets part, I only succeeded to make translations loaded in time with [a blocking XHR](https://github.com/yheuhtozr/forem/blob/d0f7a0e337e4695475216454f03c5076bfd8afcc/app/assets/javascripts/l10n.js), which apparently doesn't seem a good practice
  - in the Preact part, Webpack seems to precompile all locale files in one chunk, that means translations from all languages are loaded for every user. Is there a good way to have only file(s) corresponding to the current locale downloaded?
  - we have two lines of JS codes in this app, so I import the same i18n library twice, but is there a possibility to unify them?

Advice from JS experts would be much appreciated.

**Comment by citizen428** [comment]
I'm ok with switching libraries if there's one that's significantly better than `i18n-js`. For the rest I'll have to defer to Nick, this is not quite my area of expertise and he probably also has a better idea where we going long-term with all things frontend.

**Comment by nickytonline** [comment]
I'm in the middle of something at the moment, but I will give a think on this this week.

**Comment by yheuhtozr** [comment]
Is there any update for this?

**Comment by yheuhtozr** [comment]
@nickytonline Hi, if this needs some time, how about opening up for locale contributions in existing parts? I know that it is somewhat functionally incomplete, but many users seem to crave the feature, and there will be some hidden issues only able to capture with more eyeballs.

**Comment by Ridhwana** [comment]
> > Hm, @Ridhwana how do you feel about brining I18n to admin at this point?
> 
> I think we can hold off on the admin for the moment, we're busy reworking most of the views, which means that it's prone to big changes. As we update the views, then we can start working on them.

I want to make a note here that we're ready for I18n on the member view that we recently cleaned up on the admin. Here's an [issue](https://github.com/forem/forem/issues/16506) for it.

**Comment by Ridhwana** [comment]
> Is there any update for this?

@yheuhtozr I'll follow up on this and we'll get back to you.

**Comment by nickytonline** [comment]
I'm back from vacation @yheuhtozr. Taking a peek at this.

**Comment by nickytonline** [comment]
In regards to the YAML JSON disconnect, we could always have a build step to generate JSON from YAML and then let the frontend use that. See this demo I made at https://stackblitz.com/edit/node-85zfdo (in the terminal, run `node index.js`).

It's an extra step, but then we wouldn't need to worry about what's JSON or YAML.

For all things webpack, the localized files are json, so they would not need to be bundled via webpack. They can, but it's not necessary. I would just generate/update the json files and have them in a folder and we load the appropriate locale files based on the locale being served.

One question I have about localization in Rails and our heavily cached backend is are cachekeys now taking the locale into account as well? cc: @maestromac @Ridhwana

**Comment by yheuhtozr** [comment]
Thank you, it's good news that we will be able to load locale files without hassle. I can't say I'm very professional with the JS technology and libraries, so a proof of concept of a correctly loaded i18n library setup would be appreciated.

> It's an extra step, but then we wouldn't need to worry about what's JSON or YAML.

We could insert a small adaptation function during YAML-JSON conversion if we share strings between ruby and JS. The rails and JS (depends on library) locale file conventions are not aligned, for example:

``` yml
key:
  one: "a key"
  other: "%{count} keys"
```
``` json
{
  "key_one": "a key",
  "key_other": "{{count}} keys"
}
```
>  One question I have about localization in Rails and our heavily cached backend is are cachekeys now taking the locale into account as well?

Unfortunately I don't think so. Cache keys don't seem to carry locale information. I often see contamination from other languages on my instance, but it's maybe my mistake as I don't see something wrong on dev.to. Another issue is that few links in templates take locale into account that you basically go back to the default locale after every jump.

**Comment by aitchiss** [comment]
> it'll also mean we cannot share i18n in YAMLs directly (while I think okay because actual shared strings are small in number)

Weighing into this late, but I think this is an important point ☝️  On balance, my preference would be to pull in `i18next` for a lot of the same reasons you mention in your linked comment @yheuhtozr (the JSX support is very useful in particular), and keep the JSON/YAML separate. 

I'm wary to write conversion layers between the YAML and the JSON for the main reason being that in doing something pretty custom, we'd be increasing the complexity for people to contribute to, understand, and debug the JS portion of this functionality. Not _all_ strings are shared (I'd guess the majority are not, but I could be wrong), so I'm not convinced there's a huge benefit in keeping everything in the YAML.

`i18next` is widely used, well-supported, and I believe would be the "path of least resistance" for others in the community to contribute to this code.

cc @nickytonline

**Comment by nickytonline** [comment]
Yeah I'm definitely for i18next as well @aitchiss. The only reason I brought up the conversion was to avoid locale data in two places in the project, but it could be a trade off to maintain both YAML and JSON files. I'm thinking of this purely from a maintenance perspective as well as from a contributor point of view.

**Comment by aitchiss** [comment]
I guess my worry is the additional overhead of maintaining this conversion step from YAML -> JSON in terms of code needed to handle e.g. changing `%{count}` to `{{count}}`... 

Also, if we use HTML element placeholders like `Hello <1>{{name}}</1>` and add these to the YAML, they won't be usable by the Ruby code (AFAIK), which would be a bit confusing

**Comment by nickytonline** [comment]
That's a good point about the conversion of the actual data in the JSON being different from the YAML version. Good catch on that.

In that case, the clear separation of JSON for frontend and YAML for backend makes sense. cc: @Ridhwana

**Comment by yheuhtozr** [comment]
@nickytonline @aitchiss Hi, I'm trying to set up a configuration that asset pipeline and Webpack codes share the same locale files based on your [demo](https://stackblitz.com/edit/node-85zfdo), but struggling to make it work. Namely, when I called readFileSync it said `fs.readFileSync is not a function`. I guessed that Webpack might have not loaded node modules, so tried `environment.config.merge({target: 'node'});` in `config/webpack/environment.js` only to get `exports is not defined`. Without much experience of Webpack I cannot even figure out where is the start point to solve it.

Alternatively if I'm allowed to push my current stupid implementation with two tracks of locale JSONs, I can provide i18n for the JS part soon.

Your advices would be appreciated.

**Comment by aitchiss** [comment]
Hey @yheuhtozr! Thanks for following up on this 🙂  @nickytonline's unfortunately no longer part of the core team at Forem, having moved on to pastures new 😄 

We'll need to have a bit of an internal discussion about the approach for the JS side, taking into consideration all the great input on this issue. This might take a little bit of time, but rest assured it's on my todo list and I'll update you as soon as I can!

**Comment by yheuhtozr** [comment]
Got it, I'll wait for that as for JS.

**Comment by yheuhtozr** [comment]
> I want to make a note here that we're ready for I18n on the member view that we recently cleaned up on the admin. Here's an [issue](https://github.com/forem/forem/issues/16506) for it.

@Ridhwana Hi, could you confirm whether all admin views are basically ready for i18n or only those in `admin/users` are?

**Comment by Ridhwana** [comment]
> @Ridhwana Hi, could you confirm whether all admin views are basically ready for i18n or only those in `admin/users` are?

Hi @yheuhtozr 👋 Only recently updated views like `/admin/member_manager/users/<:id>` are ready for i18n. i18n for this page is captured in [this issue](https://github.com/forem/forem/issues/16506). 

We're planning on updating most layouts on the admin and hence we'd prefer to tackle only the newly updated views that we've worked on as the rest will most likely change in the upcoming months. 

`/admin/member_manager/users/` will be ready soon as well and I'll create an issue for it and mark it `ready for dev` once it is. 

Hope this helps give context 😊

**Comment by yheuhtozr** [comment]
Remaining i18n-related tasks in my observation:
- [ ] i18n-aware internal links
  - essential for consistent locale-based navigation (that doesn't jump back to default language UI)
  - needs to find and replace hard-coded `<a>` tags in templates
- [ ] JS frontend i18n
  - string extraction is almost done on my side
  - need to wait for the ongoing JS refactoring?
- [ ] DB model translation tables
  - needed for localization of site/user-generated contents (tag pretty names, badges, reactions, side menus...)
  - must carefully choose a library, as migration might be destructive
- [ ] LTR or writing modes support
  - is technically orthogonal to text i18n, hardly interferes with works above and can be worked concurrently
  - needs design team intervention

**Comment by mikeydorje** [comment]
Closing this out — the Hacktoberfest 2021 i18n campaign wrapped up and the bulk of the work landed.
